internal/fs/local_file_test.go

```diff
@@ -177,29 +245,135 @@ func (t *LocalFileTest) RandomWritesToLocalFile() {
+	_, f2 := t.createLocalFile("explicitFoo/" + fileName2)
+
+	// Attempt to list explicit directory.
+	entries := t.readDirectory(path.Join(mntDir, "explicitFoo/"))
+
+	// Verify entries received successfully.
+	AssertEq(2, len(entries))
+	t.verifyLocalFileEntry(entries[0], fileName1, 0)
+	t.verifyLocalFileEntry(entries[1], fileName2, 0)
+	// Close the local files.
+	t.closeFileAndValidateObjectContents(f1, "explicitFoo/"+fileName1, "")
+	t.closeFileAndValidateObjectContents(f2, "explicitFoo/"+fileName2, "")
+}
 
-	contentBytes, err := gcsutil.ReadObject(ctx, bucket, fileName)
-	AssertEq(nil, err)
-	ExpectEq(contents, string(contentBytes))
+func (t *LocalFileTest) TestReadDirForImplicitDirWithLocalFile() {
+	// Create implicit dir with 2 local files and 1 synced file.
+	AssertEq(
+		nil,
+		t.createObjects(
+			map[string]string{
+				// File
+				"implicitFoo/bar": "",
+			}))
+	fileName1 := "file1"
+	fileName2 := "file2"
+	_, f1 := t.createLocalFile("implicitFoo/" + fileName1)
+	_, f2 := t.createLocalFile("implicitFoo/" + fileName2)
+
+	// Attempt to list implicit directory.
+	entries := t.readDirectory(path.Join(mntDir, "implicitFoo/"))
+
+	// Verify entries received successfully.
+	AssertEq(3, len(entries))
+	t.verifyLocalFileEntry(entries[0], "bar", 0)
+	t.verifyLocalFileEntry(entries[1], fileName1, 0)
+	t.verifyLocalFileEntry(entries[2], fileName2, 0)
+	// Close the local files.
+	t.closeFileAndValidateObjectContents(f1, "implicitFoo/"+fileName1, "")
+	t.closeFileAndValidateObjectContents(f2, "implicitFoo/"+fileName2, "")
 }
 
-func (t *LocalFileTest) createLocalFile(fileName string) (filePath string, f *os.File) {
-	// Creating a file shouldn't create file on GCS.
-	filePath = path.Join(mntDir, fileName)
-	f, err := os.Create(filePath)
+func (t *LocalFileTest) TestRecursiveListingWithLocalFiles() {
+	// Create sub-directory.
+	AssertEq(
+		nil,
+		t.createObjects(
+			map[string]string{
+				"subDir/": "",
+			}))
+	// Create implicit dir with 1 local file1 and 1 synced file.
+	AssertEq(
+		nil,
+		t.createObjects(
+			map[string]string{
+				// File
+				"subDir/implicitFoo/bar": "",
+			}))
+	fileName1 := "implicitLocalFile"
+	_, f1 := t.createLocalFile("subDir/implicitFoo/" + fileName1)
+	// Create explicit dir with 1 local file.
+	AssertEq(
+		nil,
+		t.createObjects(
+			map[string]string{
+				"subDir/explicitFoo/": "",
+			}))
+	fileName2 := "explicitLocalFile"
+	_, f2 := t.createLocalFile("subDir/explicitFoo/" + fileName2)
+	// Create local file in mnt/SubDir dir.
+	fileName3 := "baseLocalFile"
+	_, f3 := t.createLocalFile("subDir/" + fileName3)
 
-	AssertEq(nil, err)
-	t.validateObjectNotFoundErr(fileName)
+	// Recursively list mntDir/subDir/ directory.
+	cmd := exec.Command("ls", "-R", path.Join(mntDir, "subDir/"))
+	out, err := cmd.Output()
 
-	return
+	// Verify output.
+	AssertEq(nil, err)
+	AssertTrue(strings.Contains(string(out), "/subDir:\nbaseLocalFile\nexplicitFoo\nimplicitFoo"))
```

Comment:
can you try using filepath.walk or filepath.walkDir to recursively all files under directory.  Creating one cache for all buckets
Refactoring the code to use same file-cache implementation

We can avoid assertion on a string